src/ast/passes/args_resolver.cpp

```diff
@@ -0,0 +1,118 @@
+#include <bpf/bpf.h>
+#include <cassert>
+
+#include "arch/arch.h"
+#include "ast/passes/args_resolver.h"
+#include "ast/passes/probe_expansion.h"
+#include "ast/visitor.h"
+#include "bpftrace.h"
+#include "dwarf_parser.h"
+#include "probe_matcher.h"
+#include "tracepoint_format_parser.h"
+#include "util/strings.h"
+
+namespace bpftrace::ast {
+
+namespace {
+
+class ArgsResolver : public Visitor<ArgsResolver> {
+public:
+  explicit ArgsResolver(ASTContext &ast,
+                        BPFtrace &bpftrace,
+                        ExpansionResult &ap_expansions,
+                        std::unordered_set<ProbeType> probe_types)
+      : ast_(ast),
+        bpftrace_(bpftrace),
+        ap_expansions_(ap_expansions),
+        probe_types_(std::move(probe_types))
+  {
+  }
+
+  using Visitor<ArgsResolver>::visit;
+  void visit(Builtin &builtin);
+  void visit(Probe &probe);
+
+private:
+  void resolve_args(Probe &probe);
+
+  ASTContext &ast_;
+  BPFtrace &bpftrace_;
+  ExpansionResult &ap_expansions_;
+  std::unordered_set<ProbeType> probe_types_;
+  Probe *probe_ = nullptr;
+};
+
+} // namespace
+
+void ArgsResolver::visit(Builtin &builtin)
+{
+  if (builtin.ident == "args" || builtin.ident == "__builtin_retval") {
+    resolve_args(*probe_);
```

Comment:
Isn't this now redunant with the field_analyser? I would add `ctx` in here, and then strip that functionality from the field analyser. The field analyser would now be exclusively used for resolving the non-BTF cast types?

Reply:
Right, can we just pull that tracepoint arg expansion into this pass? I think it makes sense since this is `args_resolver`?

Reply:
Gonna wait for @viktormalik to do the tracepoint args refactoring before touching any more of this.

Reply:
Yeah, I'll see how it changes with the refactoring that I'm working on.